### ISSUE TITLE
Provide a way to extend the chain that just mirrors the first part.

### ISSUE THREAD

**Issue by marcmerlin** [hzeller/rpi-rgb-led-matrix#1000]
I have a panel setup where my display is mirrored onto a 2nd display, so I use ribbon cables to plug the 2nd set of panels as output of the first ones.
I don't get or want a wider display, I just use the fact that the pixels get shifted out to load the 2nd panels.
This works fine, except that with FM2616A, the init sequence doesn't get sent wider than the display size, so the mirrored display doesn't get initialized.

Workaround: send a bogus frame that is twice as wide to init all the panels
Better workaround, some init option to give a width init factor (default to 1, but could be 2, 3, whatever)

**Comment by marcmerlin** [comment]
Also, 1000! :)

**Comment by hzeller** [comment]
mmh, usually if you add another panel beyond the length, the output is very poor there: the bits are shifted there, but the timing of course does not line up (the bits are off-by-one), so colors are horrible.
So to mirror, you also have to send the data twice.

Not sure how useful such feature is given that the can just declare a larger frame and set the pixels twice.

**Comment by marcmerlin** [comment]
Let me test that, I've done that with the smartmatrix driver, and it works perfectly fine.
I understand however that it depends how you shift bits (at a given interval no matter what, or if you use BCM which means that the shifted bits will stay the wrong amount of time on the 2nd panel.

Here's a smartmatrix example where the panels are daisy chained and only 64x64 is sent to 192x64 (i.e. the same pattern is shifted twice and looks exactly the same on the secondary panels)
![image](https://user-images.githubusercontent.com/1369412/76460138-b6ba5680-639a-11ea-8ef9-a6562b876b23.png)

I'll try with your driver to see how it goes

**Comment by marcmerlin** [comment]
@hzeller Check out https://photos.app.goo.gl/iH1BeCyrtAKFfLBS8 for how it looks like.

I refreshed my memory with https://github.com/hzeller/rpi-rgb-led-matrix/issues/947
The trick is that if you set FM6126A on panels that already work, and only send 128 pixels deep, it will break the init on the other panels in parallel and shut them off.
So I sent them the init string, then changed chain length to 1 and removed FM6126A, and got it to work.
I'm pasting the worst one below where you can tell the colors are not the same, but still close enough for me, and given that I'm doing color demos, having the exact right color isn't important anyway
![image](https://user-images.githubusercontent.com/1369412/76461826-dacb6700-639d-11ea-9f3e-fdfeac18e9de.png)

Long story short, you are correct that the colors are not exactly the same, but they are close enough that it doesn't really matter for my use. I'm not sure how you would send the data twice without splicing cables. Is there a good way to do this?
Otherwise, if I end up using my hacked up solution, I'm fine patching your library for my own use, I understand if you don't really want to add/support an option that apparently no one but me, will use.

I think I already sent you this, but in case you were wondering why the need for mirror display, it's for this use. I have the same display front and back, and to ease wiring, I simply shift the front panels into the rear ones.
![image](https://user-images.githubusercontent.com/1369412/76462060-43b2df00-639e-11ea-9626-6a2ac52a8431.png)

**Comment by marcmerlin** [comment]
And basically my previous comment is saying "feel free to close this issue as WONTFIX" if you don't think there is a reasonable need to support this (there probably isn't), and I'll patch my way around it (will probably send the patch to this issue if anyone finds it later with google and wants it)

**Comment by hzeller** [comment]
Changed this to feature request. Essentially we want a `--led-repeat-chain`.

**Comment by andytheengineerguy** [comment]
Hi @hzeller and @marcmerlin, I know this thread is from a while ago, but I'm curious, I am now wanting to do the same thing, and was wondering if there were any updates or workarounds found for this in the last 3 years?

Here's how far I got with my attempts tonight: I have a square 128x128 panel (2 x 64x128) that I want to mirror. Similar to you Marc, I'm displaying these as screens on the front and back of an outfit, so I want to mirror and not extend these squares (extending ends up being too many pixels for the CPU to display nicely) 

Tonight I made some custom daisy-chained ribbon cables to try and achieve this goal:

<img width="1146" alt="image" src="https://github.com/hzeller/rpi-rgb-led-matrix/assets/26783020/dc342c5e-a599-4296-a181-496efb3d3c56">


This results in:
- Chain 1 is sending out a 64x128 signal to the two top panels
- Chain 2 is sending out a 64x128 signal to the two bottom panels

To my surprise, this actually worked pretty well. The top panels of the screens (chain 1) create an almost identical duplicate (even the colors look the same). However, the bottom panels of the screens (chain 2), display some tearing/ghosting/blinking/imperfections. It's hard to show this with a static image, but the area in the circle is aggressively flickering.

<img width="1005" alt="image" src="https://github.com/hzeller/rpi-rgb-led-matrix/assets/26783020/db4e055a-2a8b-4ae7-a815-9c4cc7085941">


Any idea why this might be happening, or what steps I could take to fix it? Do the ribbon cables need to be more of a 'Y' split cable, rather than a single ribbon with a drop-off point in the middle? Any suggestions welcome, I'm pretty new to this level of LED debugging! Thanks.

**Comment by andytheengineerguy** [comment]
ha, right after I posted this, I experimented with changing the gpio slowdown, and seemed to have stumbled into a solution!

I was previously slowing it down to 3. Then set the value to 4 and these imperfections magically disappeared.

Out of curiousity.. Any idea why slowing down the GPIOs helped with this issue? If not, no probs, I'm happy with this as it is!

**Comment by hzeller** [comment]
Yes, slowing down helps.

Also, making the data cables as short as possible is probably good, but I guess given the size of the panels, that might not be possible. There can be weird reflections at the end of the cables, so playing with a y-configuration might be interesting. You use an active adapter as an interface between your Pi and the boards, right ?

Another thing to look out for is that the panels have a solid power supply. I have seen panels glitch in the way you describe if they didn't get full 5V, or the cable to the power supply was too long. Sometimes a low-ESR capacitor at each power input at the panels can help as the panels draw very spiky load from the power supply (There are some musing about this in the [a word about power](https://github.com/hzeller/rpi-rgb-led-matrix/blob/master/wiring.md#a-word-about-power) section).

**Comment by ledvinap** [comment]
@andytheengineerguy : Another possible fix is to use two active boards, connected to RPI in parallel.

**Comment by marcmerlin** [comment]
@ledvinap  I already use active board with 3 outputs to get my 128x192 at max speed, and then divide speed by 2 since I shift pixels from the first display to the 2nd one for mirroring.
It would still be awesome if the driver could change the order it does BCM (I'm getting from LSB to MSB or the other way around) so that if you daisy chain it actually mostly looks ok without having to send the pixels twice, as per the smartmatrix example I gave above, but in the meantime, mirroring works for me, and I don't have a signal loss issue since the signal is reboosted from the 1st panel to the 2nd one and my ribbon cables aren't too long.

**Comment by ledvinap** [comment]
@marcmerlin : IMO, when connected this way, you get bitplanes shifted one one PWM period on second panel. It will always ruin colors. 
Using 2x active boards stacked to RPI GPIO shall fix signal integrity issues, with identical image on both displays.

**Comment by andytheengineerguy** [comment]
hi @hzeller, thanks for your response. I am using an active-3 board from electrodragon yes. I might try with Y-ribbon cables, if I do I'll post results here. I indeed read your section on power and was considering experimenting with the low-ESR capacitors. The final product will be battery driven, so I'm motivated to reduce the current spikes regardless, so thanks, I'll give this a shot 💪

@marcmerlin I'm not understanding your approach just yet; if you simply want to mirror/duplicate your display, can you not just connect the data bus in parallel using custom cables as I have above (keeping all else the same)? What is the reason behind dividing the speed by 2 and then sending pixels twice?

@ledvinap I'm also not 100% sure I understand this too. Are you recommending using 1 active board per LED chain?

**Comment by marcmerlin** [comment]
@andytheengineerguy, the reason for my approach simply ease and length of cabling. It's easier to take the hub75 out connectors on the front panels and connect them to the hub75 in connectors of the rear panels
![image](https://github.com/hzeller/rpi-rgb-led-matrix/assets/1369412/9bd024d8-a1ff-4323-9368-e35dfd3e0149)

I run 3 busses in parallel, but indeed I have to push the pixels twice for each frame, so each panel runs at half speed, but since I'm only pushing 256x64 per bus, it's still fast enough.

If I had easy to find Y cables of the right length, I would consider them, it would double my refresh rate. I'm just not equipped to make my own.

**Comment by andytheengineerguy** [comment]
@marcmerlin ahhh ok, that makes sense. Given that the refresh rate is enough for your needs, it's a good call to simplify the wiring by just pushing the pixels twice / frame. Btw, if you're interested, you can use this [IDC crimp tool](https://www.amazon.com/gp/product/B072MM8X7Y/ref=ppx_yo_dt_b_search_asin_title?ie=UTF8&psc=1) to easily make your own cables, I bought this recently and it works great 👌

Regarding the active board chains, I see what you're saying, but for the sake of keeping the 128*128 screens 100% equal I think I'll stick with what I'm already doing, which is using 1 chain for the top 2 panels and the other chain for the bottom 2 panels. Each chain has its own logic buffer and 1 chain will be remain unused, but I seem to be getting high enough refresh rates anyway (~240).

**Comment by marcmerlin** [comment]
@andytheengineerguy  totally for 128x128, 2 chains is more than plenty. And if you tweak some colors/precision settings, you should be able to get over 500Hz :)

**Comment by andytheengineerguy** [comment]
Hmmm I'm tryna retain as much color accuracy as I can... So even with the effectively doubled speed from these parallelized ribbon cable solution, plus dither_bits =1, plus reducing the lsb_nanoseconds to 90, I still need to reduce the PWM bits from 11 down to 8 to get ~230Hz...

Kinda off topic of the thread now, but this is my config and I'm still curious to ask for a 2nd opinion as I've been working mostly in a silo so far:
<img width="361" alt="image" src="https://github.com/hzeller/rpi-rgb-led-matrix/assets/26783020/8a86fe76-3f9c-4184-b448-e71b4ef1e09a">

I also found that the python bindings performed almost exactly the same as C. This was only comparing the gif and image viewer examples. Eg, using these same settings with the C image viewer + the same gif also gives me 230-240 Hz, qhich is equal to this python script.

Is this your experience too?

**Comment by marcmerlin** [comment]
Python bindings are just bindings, the actual code still runs in C++ at full speed, so you should get the same refresh rate.
I care about Hz more than perfect color, so mine is tuned a bit differently:
https://github.com/marcmerlin/FastLED_NeoMatrix_SmartMatrix_LEDMatrix_GFX_Demos/blob/42dc695dcf49369099f430100a9bf8f88453d0ca/neomatrix_config.h#L1321

**Comment by andytheengineerguy** [comment]
I was always under the impression that anything above ~250Hz was diminishing returns; our eyes are good at ~100 and phone cameras never work well with this highly multiplexed scan architecture that comes with screens like ours (assuming yours is 1/64 or 1/32?). whats the reason for aiming for such high Hz? Are you filming in slow motion or something?

**Comment by marcmerlin** [comment]
eyes don't need it, cameras can take shots at 0.001s and if so, they'll see the refresh bars, or you even see some black or other issues.
Having a much faster refresh rate than the camera helps with this (I'm 1/32).
I have plenty of pictures of my outfit taken with cellphones that I had to throw away
![image](https://github.com/hzeller/rpi-rgb-led-matrix/assets/1369412/51a02d63-2e1f-4502-9f20-977924285078)

**Comment by andytheengineerguy** [comment]
Ahh I was wondering - I'm also creating some LED art to show at music festivals, and yeah, want to be able to take pictures with it. But I'm finding that we're limited by the architecture more than anything else.... This is a screenshot of 3 pics I took @ 470Hz, absolutely bonkers FPS under normal circumstances, but apparently still not good enough when only 1/64th of the screen is illuminated at once. Even when the black lines don't appear, there are noticeable white lines that make it look weird.

![image](https://github.com/hzeller/rpi-rgb-led-matrix/assets/26783020/885e4b86-8f33-4aed-8593-b6690e5e7117)

I think thats why I said the defeated-sounding "dimishing returns" comment haha - if 500Hz doesn't cut it, what will. A DSLR camera maybe?

**Comment by marcmerlin** [comment]
Mmmh, that's disappointing indeed :-/ 
Note that a DSLR in auto mode may be worse because if it can take a lot of light, it will snap the picture very quickly and could catch the bars. If it's darker or the sensor isn't as good, it will need more time for a shot and the bars will go away

**Comment by andytheengineerguy** [comment]
Just purchased some 1/16 scan rate, and also a less common 1/13 scan rate to experiment with. Will see how it goes. Happy to share results if you're interested.

**Comment by marcmerlin** [comment]
Sure thing, thanks

**Comment by marcmerlin** [comment]
With apologies to everyone, this project is doing a one time cleanup of 
all the bugs that have not been addressed and closed in the last 10 years.

If you filed an issue to ask for help, please do not re-open this one or a new
bug, but go here instead: https://rpi-rgb-led-matrix.discourse.group/

Please be mindful that this project does not have any full time maintainers
or anyone paid. You should not expect personal help for your own projects or
answers to programming questions but you can ask for help with panels that
aren't working, setup questions, or anything that may be a bug here:
https://rpi-rgb-led-matrix.discourse.group/
If you need programming help and are willing to pay someone for it, please
state so in your message, and someone may be interested.

If you did report an actual bug that is still not fixed today, please accept
our apologies for this being mass closed, you can absolutely send a followup
and re-open the issue, or open a new up to date one as applicable.
Note that we very much welcome help, so if you can help fix your bug, you will
be a superstar :)

If your bug was linked to a pull request that was not resolved, please update
the pull request and the issue, and we will do our best to review them. We do
thank you for your undertanding that we have very limited free time, so any
extra work you do along with your bug report or PR request will be very much
appreciated and improve the odds that they will be looked at.

If you want to help contribute as a python or C# or any other language bindings,
you are also a superstar, please state so in the issue and we'll do our best to
give you the access and reviews you need.

For anything related to PWM panel support, please make sure you read README and
then refer to https://github.com/hzeller/rpi-rgb-led-matrix/issues/466 . 
Please do not comment if you're haven't read everything and/or if you're just
asking when your panel is going to be supported. Those bugs are if you're 
able to contribute to the effort.

For rPI5: https://github.com/hzeller/rpi-rgb-led-matrix/issues/1603

For anything python, please link your bug to: https://github.com/hzeller/rpi-rgb-led-matrix/issues/1828
For anything C#, please link your bug to: https://github.com/hzeller/rpi-rgb-led-matrix/issues/1828
